docs/kuneiform/procedures.md

```diff
@@ -0,0 +1,275 @@
+
+## Procedure Blocks
+
+Procedures are declared using the `procedure` key word. They can be given a list of parameters and their corresponding types, return types, and access modifiers. The example below shows the signature for a `get_user` procedure. It takes one input (the id of the user, of type `uuid`), and returns the user's name and age. It will return exactly one record, containing a name and age:
+
+```ts
+procedure get_user($id uuid) public view returns (name text, age int) {
+    // procedure implementation
+}
+```
+
+Procedures can also return tables, which are a collection of many records. The below example returns all user IDs and names of a certain age:
+
+```ts
+procedure get_users($age int) public view returns table(id uuid, name text) {
+    // procedure implementation
+}
+```
+
+If a procedure does not return anything, the return should not be specified:
+
+```ts
+procedure create_user($name text, $age int) public {
+    // procedure implementation
+}
+```
+
+## Procedure Syntax
+
+### Declaring and Assigning Variables
+
+All variables in procedures are strongly typed. They can be assigned at declaration, or assigned to after declaration. They can also be type casted:
+
+```ts
+procedure declare_vars() public {
+    // declare and assign new int var:
+    $var1 int := 1;
+
+    // declare a new var to be assigned to later
+    $var2 text;
+
+    // assign to $var2. type cast to match type text:
+    $var2 := $var1::text;
+}
+```
+
+### Arithmetic
+
+Basic arithmetic can be performed in procedures:
+
+```ts
+procedure do_math($num1 int, $num2 int) public {
+    // add and divide, with precedence:
+    $avg int := ($num1 + $num2)/2;
+
+    // modulo:
+    $mod int := $num2%10;
+
+    // negate
+    $neg_mul int := (-$avg)*$mod;
+}
+```
+
+### Executing SQL
+
+SQL statements can be executed at any point in a procedure. In the below example, we generated a deterministic UUID using a random UUID namespace and the transaction ID. We then use the generated id as the user's primary key.
+
+```ts
+procedure create_user($name text, $age int) public {
+    $id uuid := generate_uuid_v5('455f60aa-0569-4aaa-8469-63be2ec4dd96'::uuid, @txid)
+    // do something
+    // ...
+    INSERT INTO users (id, name, age)
+    VALUES ($id, $name, $age);
+}
+```
+
+### Handling Arrays
+
+Procedures support basic array operations. It supports the following functions:
+
+- `array_append(type[], type) type[]`: appends a value to the end of an array.
+- `array_prepend(type, type[]) type[]`: prepends a value to the beginning of an array.
+- `array_cat(type[], type[]) type[]`: concatenates two arrays together.
+- `array_length(type[]) int`: returns the length of the array.
+
+**Arrays are 1-indexed.**
+
+```ts
+procedure handle_array() public view returns (val text) {
+    $arr text[] := ['hello', 'world'];
```

Comment:
Mostly unrelated to this `handle_array` example, but do we support fixed length arrays like `int[16]` as postgres does, or would that have no value in procedures or procedure returns?

Reply:
We currently do not support that